File: 0003/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        first_ptr = 0
        second_ptr = 0
        
        longest = 0
        chars = dict()
        
        while second_ptr < len(s):
            curr = s[second_ptr]
            
            if curr in chars:
                longest = max(longest, len(chars))
                idx_curr = chars[curr]
                
                while first_ptr < idx_curr:
                    del chars[s[first_ptr]]
                    first_ptr += 1
                
                first_ptr += 1
            
            chars[curr] = second_ptr
            second_ptr += 1
                        
        return max(longest, len(chars))

if __name__ == "__main__":
    assert(Solution().lengthOfLongestSubstring("abcabcbb") == 3)
    assert(Solution().lengthOfLongestSubstring("bbbbb") == 1)
    assert(Solution().lengthOfLongestSubstring("pwwkew") == 3)

File: 0003/main.rs
struct Solution { }

use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut start_iter = s.chars();
        let mut start: usize = 0;
        
        let mut longest: usize = 0;
        let mut map: HashMap<char, usize> = Default::default();
        
        for (idx, curr) in s.chars().enumerate() {
            if map.contains_key(&curr) {
                if map.len() > longest {
                    longest = map.len();
                }
                
                let idx_curr = *map.get(&curr).unwrap();
                while start < idx_curr {
                    map.remove(&start_iter.next().unwrap());
                    start += 1;
                }
                
                start_iter.next();
                start += 1;
            }
            
            map.insert(curr, idx);
        }
        
        if map.len() > longest {
            longest = map.len();
        }
        
        return longest as i32;
    }
}

fn main() {
    assert_eq!(Solution::length_of_longest_substring("abcabcbb".into()), 3);
    assert_eq!(Solution::length_of_longest_substring("bbbbb".into()), 1);
    assert_eq!(Solution::length_of_longest_substring("pwwkew".into()), 3);
}